### DISCUSSION TITLE
3.1.0

### DISCUSSION THREAD

**Discussion by janusw** [dotnet-bluetooth-le/dotnet-bluetooth-le#859] [Announcements]
## What's Changed
(relative to 3.0.0)
* Make the Maui sample client work on Windows by @janusw in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/758
* Some package updates in Plugin.BLE.Tests and BLE.Client by @janusw in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/759
* Windows.Devices.Bluetooth Only by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/746
* ReadAsync updates characteristic value on Windows by @Praseodymi in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/764
* Remove .NET 6 by @janusw in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/770
* Added BroadcastName to AdvertisementRecordType by @janmarektelc in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/773
* disable MainThreadInvoker in Adapter.ConnectToDeviceAsync by @janusw in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/774
* StateChanged for Windows by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/777
* Add support for .NET 8 by @janusw in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/778
* Improve Connect In Windows by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/776
* GHA: add a Linux build by @janusw in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/791
* Turn Bluetooth Radio On/Off by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/784
* Connection parameters for windows by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/801
* GHA: build Maui client on Windows by @janusw in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/806
* Added services for citysports treadmill by @hughesjs in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/805
* Fix GetSystemConnectedOrPairedDevices in Windows by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/807
* Fix ConnectionLost -> Connect for Windows by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/809
* Fix some warnings by @janusw in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/811
* Implement BleImplementation.TrySetStateAsync on Android by @janusw in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/823
* Fix a NullReferenceException in Characteristic.NSErrorToGattStatus by @janusw in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/827
* Windows Implementation for BondAsync and BondedDevices by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/820
* Maui client update by @jeetee in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/837
* Windows: Handle that a device is already connected by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/839
* fix Device.UpdateRssiAsync on Windows by @janusw in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/844
* Implement Adapter.SupportsExtendedAdvertising on Windows by @janusw in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/845
* Windows: Fixed null entries in DiscoveredDevices  #828 by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/848
* Windows: Connect -> ConnectionLost -> Reconnect by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/849
* Windows: Check for OS build number version Runtime by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/853
* fix #854 - Bluetooth stack breaks when trying connect on phone boot by @alex3696 in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/855
* Fixed StartUpdatesNativeAsync for indicate by @AskBojesen in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/858

## New Contributors
* @AskBojesen made their first contribution in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/746
* @Praseodymi made their first contribution in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/764
* @janmarektelc made their first contribution in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/773
* @hughesjs made their first contribution in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/805
* @jeetee made their first contribution in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/837
* @alex3696 made their first contribution in https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/855

**Full Changelog**: https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/compare/3.0.0...3.1.0

<hr /><em>This discussion was created from the release <a href='https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/releases/tag/3.1.0'>3.1.0</a>.</em>

**Comment by AskBojesen** [comment]
@janusw : I have a suggestion for the version numbering of the next release.
My problem with the current versioning scheme is that the assembly version and file version has been 3.1.0.0 constantly during the development for the Plugin.BLE.dll

The Windows installer I use for tools differentiates only using the file version, so an update does not update the dll.
Can you maybe increase the patch number during development and release, so they in fact get a different file version?
Eg. 3.1.0.234

![image](https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/assets/131768758/030278bf-354b-46c2-b3ec-768a7f4ebf3b)

**Reply by janusw** [reply]
> @janusw : I have a suggestion for the version numbering of the next release. My problem with the current versioning scheme is that the assembly version and file version has been 3.1.0.0 constantly during the development for the Plugin.BLE.dll
> 
> The Windows installer I use for tools differentiates only using the file version, so an update does not update the dll.

I was not aware of this problem.

What we have been doing so far is this:
* The `<Version>` in the csproj  files is only updated for tagged versions (e.g. `3.1.0-beta.2` or `3.1.0-rc.1`), but not in between.
* The version of the nupkg files built by GitHub Actions is set via `git describe` (something like `3.1.0-rc.1-19-g1e19fa8`).

I think the `<FileVersion>` and `<AssemblyVersion>` are set from the `<Version>`, but their format is more restricted (`major.minor.patch.build`). Since we're only using `major.minor.patch` in the tags, the `build` number is always 0 this way and we  end up with 3.0.0.0 or 3.1.0.0.

> Can you maybe increase the patch number during development and release, so they in fact get a different file version? Eg. 3.1.0.234

Well, I guess we don't want to set the `<FileVersion>` and/or `<AssemblyVersion>` manually in each commit, but it should e.g. be possible to set them automatically in GHA.

Do you have any suggestions how to best handle this in a somewhat automated fashion?

**Reply by jeetee** [reply]
A possible approach I'm trying out on some projects is to use the build datetime as FileVersion. Assembly Version then stays as per release numbering.

`
<FileVersion>$([System.DateTime]::UtcNow.ToString('yyyy.MM.dd')).$([System.Convert]::ToUInt16($([MSBuild]::Divide($([System.DateTime]::UtcNow.TimeOfDay.TotalSeconds), 2))))</FileVersion>`

Which'll lead to a FileVersion looking like 2024.5.23.39408 for example; the upside (in my very humble opinion) is that you can use FileVersion to always identify a newer vs older file (if builds differ at least 2s).

**Reply by janusw** [reply]
> A possible approach I'm trying out on some projects is to use the build datetime as FileVersion. Assembly Version then stays as per release numbering. [..]
> Which'll lead to a FileVersion looking like 2024.5.23.39408 for example; the upside (in my very humble opinion) is that you can use FileVersion to always identify a newer vs older file (if builds differ at least 2s).

Well yes, but that only tells you something about the build date. You'd have no idea if that build was from current master (3.1.0+) or from some old branch (e.g. v3.0.0 or v2.1.3).

My idea would rather be to use `major.minor.patch` from the tags, as it is currently done, and then maybe have the `build` number as the number of commits since the last tag (or the last stable/annotated tag, to be precise).

So the third commit after the tag `3.1.0` would have the version `3.1.0.3`. That still does not tell you whether the commit is on the master branch or some other experimental branch, but at least you get a rough idea where you are in the versioning scheme, and all commits on master would have a unique and increasing number.

**Reply by jeetee** [reply]
True.. We use FileVersion indeed as only that.
Then we make use of `VersionPrefix` (which can be provided via build parameters) to contain Major.Minor.Patch (based of the tag) and `VersionSuffix`, (defaults to 'dev' for local builds, could indeed be your proposed commit count).

We configure `AssemblyVersion` to match `VersionPrefix` to avoid it expanding with the .0 revision number. So in our usage AssemblyVersion is the main indicator and FileVersion is used as a pure build indicator.

But for the purpose here I'd then propose to configure `AssemblyVersion` as `Version` so it'd retain the counting number from the suffix

```
<!-- The prefix contains jist Major.Minor.Build information -->
<VersionPrefix Condition="'$(VersionPrefix)' == ''">0.1.0429</VersionPrefix>
<!-- We use a default suffix to indicate development builds if no suffix was set -->
<VersionSuffix Condition="'$(VersionSuffix)' == ''">dev</VersionSuffix>
<!-- The version number actually used for assembly matching/linking
	For libraries: Should only change when breaking compatibility
	For applications: Should change each release -->
<AssemblyVersion>$(Version)</AssemblyVersion>
<!-- Using Build DateTime as Assembly File Version, Revision is seconds/2 to fit within uint16 -->
<FileVersion>$([System.DateTime]::UtcNow.ToString('yyyy.MM.dd')).$([System.Convert]::ToUInt16($([MSBuild]::Divide($([System.DateTime]::UtcNow.TimeOfDay.TotalSeconds), 2))))</FileVersion>
<!-- Don't set InformationVersion, it'll default to $(Version) with $(SourceRevisionId) appended to it. This is the field visible as Product Version in the Detailed File Properties window in Windows Explorer -->
```

**Comment by axa88** [comment]
@AskBojesen @janusw 

Ive installed this latest 3.1.0 as confirmed in VS:
![image](https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/assets/13371986/10a85d0b-99fc-4b97-8340-3cf0dff5a5de)

Ive cleaned the solution, deleted the `obj` and `bin` folders, rebuilt, and when i attempt to BOND on windows Im getting an exception due to this comparison of the `NativeDeivice` in `Source/Plugin.BLE/Windows/Adapter.cs`


```
public override async Task BondAsync(IDevice device)
{
	if (!(device.NativeDevice is BluetoothLEDevice bluetoothLEDevice))
	{
		Trace.Message($"BondAsync failed since NativeDevice is null with: {device.Name}: {device.Id} ");
		return;
	}
   ...
}
```

But as far as I see that is invalid, and the `master` and `3.1.0` tag repos  both have:

```
public override async Task BondAsync(IDevice device)
        {
            var bleDevice = device.NativeDevice as BluetoothLEDevice;
            if (bleDevice is null)
            {
                Trace.Message($"BondAsync failed since NativeDevice is null with: {device.Name}: {device.Id} ");
                return;
            }
    ...
}
```

Could something have been built wrong in the `3.1.0` nuget?
Does the version of windows targeted matter?
Im about to throw my pc out the window... any help is appreciated.

**Reply by janusw** [reply]
> Ive cleaned the solution, deleted the `obj` and `bin` folders, rebuilt, and when i attempt to BOND on windows Im getting an exception due to this comparison of the `NativeDeivice` in `Source/Plugin.BLE/Windows/Adapter.cs`
> 
> ```
> public override async Task BondAsync(IDevice device)
> {
> 	if (!(device.NativeDevice is BluetoothLEDevice bluetoothLEDevice))
> 	{
> 		Trace.Message($"BondAsync failed since NativeDevice is null with: {device.Name}: {device.Id} ");
> 		return;
> 	}
>    ...
> }
> ```

Where does the code that you quote here even come from? It's not in the repo.



> But as far as I see that is invalid, and the `master` and `3.1.0` tag repos both have:
> 
> ```
> public override async Task BondAsync(IDevice device)
>         {
>             var bleDevice = device.NativeDevice as BluetoothLEDevice;
>             if (bleDevice is null)
>             {
>                 Trace.Message($"BondAsync failed since NativeDevice is null with: {device.Name}: {device.Id} ");
>                 return;
>             }
>     ...
> }
> ```

Yes, this was added in #820 (before 3.1.0-rc.1).

> Could something have been built wrong in the `3.1.0` nuget?

I doubt it. Instead, could it be that you local package cache is somehow messed up? Have you possibly built a package yourself with version 3.1.0 that is still lingering in the package cache?


> Does the version of windows targeted matter?

In general: Yes, it matters. But what are you asking specifically? Does it matter for what?
